src/command/backup/backup.c

```diff
@@ -0,0 +1,1906 @@
+        FUNCTION_TEST_PARAM(INT, queueIdx);
+        FUNCTION_TEST_PARAM(UINT, queueTotal);
+    FUNCTION_TEST_END();
+
+    // Move (forward or back) to the next queue
+    queueIdx += clientIdx % 2 ? -1 : 1;
+
+    // Deal with wrapping on either end
+    if (queueIdx < 0)
+        FUNCTION_TEST_RETURN((int)queueTotal - 1);
+    else if (queueIdx == (int)queueTotal)
+        FUNCTION_TEST_RETURN(0);
+
+    FUNCTION_TEST_RETURN(queueIdx);
+}
+
+// Callback to fetch backup jobs for the parallel executor
+typedef struct BackupJobData
+{
+    const String *const backupLabel;                                // Backup label (defines the backup path)
+    const bool backupStandby;                                       // Backup from standby
+    const String *const cipherSubPass;                              // Passphrase used to encrypt files in the backup
+    const bool compress;                                            // Is the backup compressed?
+    const unsigned int compressLevel;                               // Compress level if backup is compressed
+    const bool delta;                                               // Is this a checksum delta backup?
+    const uint64_t lsnStart;                                        // Starting lsn for the backup
+
+    List *queueList;                                                // List of processing queues
+} BackupJobData;
+
+static ProtocolParallelJob *backupJobCallback(void *data, unsigned int clientIdx)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM_P(VOID, data);
+        FUNCTION_TEST_PARAM(UINT, clientIdx);
+    FUNCTION_TEST_END();
+
+    ASSERT(data != NULL);
+
+    ProtocolParallelJob *result = NULL;
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // Get a new job if there are any left
+        BackupJobData *jobData = data;
+
+        // Determine where to begin scanning the queue (we'll stop when we get back here).  When copying from the primary during
+        // backup from standby only queue 0 will be used.
+        unsigned int queueOffset = jobData->backupStandby && clientIdx > 0 ? 1 : 0;
+        int queueIdx = jobData->backupStandby && clientIdx == 0 ?
+            0 : (int)(clientIdx % (lstSize(jobData->queueList) - queueOffset));
+        int queueEnd = queueIdx;
+
+        do
+        {
+            List *queue = *(List **)lstGet(jobData->queueList, (unsigned int)queueIdx + queueOffset);
+
+            if (lstSize(queue) > 0)
+            {
+                const ManifestFile *file = *(ManifestFile **)lstGet(queue, 0);
+
+                // Create backup job
+                ProtocolCommand *command = protocolCommandNew(PROTOCOL_COMMAND_BACKUP_FILE_STR);
+
+                protocolCommandParamAdd(command, VARSTR(manifestPathPg(file->name)));
+                protocolCommandParamAdd(
+                    command, VARBOOL(!strEq(file->name, STRDEF(MANIFEST_TARGET_PGDATA "/" PG_PATH_GLOBAL "/" PG_FILE_PGCONTROL))));
+                
```

Comment:
Why is this check for determining to save a copy not inside the job loop above? If we go through 100 jobs and at the 50th job the sizeCopied is already > manifestSaveSize, we won't save until we've completed the 100th job.

Reply:
You are right, but it doesn't seem worth doing this in the loop since we are only logging.  We'll save the manifest before we go wait for more completed jobs.